### PR TITLE
[Test] Add reusable device redzones with corruption injection coverage

### PR DESCRIPTION
**Pull request by 0z5a**
## Scope

Related to https://github.com/flashinfer-ai/flashinfer/issues/3978 — this PR takes the portable red-zone / canary sub-scope only: a reusable test-side helper, safe fault injection, and one real integration. It does not attempt to reproduce the original SM100 issue, and it does not build the asset-snapshot / scattered-sentinel system.

In scope: guarded allocation with an owning storage and a payload view, byte-wise guard verification with region/offset reporting, corruption injection that stays inside the owning allocation, and a real pre-allocated-output call site.
Out of scope: any new public production API, non-contiguous / overlapping-stride layouts, cross-device corruption claims.

## Implementation

- One single owning allocation per buffer: `[prefix guard 0xA5][alignment pad 0x00][payload][suffix guard 0x5A][alignment slack]`, with `payload.data_ptr() % alignment == 0` by construction and the payload a view of that same storage, so the owner cannot outlive nothing and no dangling pointer is handed to a kernel.
- Guards are compared as **bytes**, never through float equality — a NaN payload cannot produce a false positive — and only the two initialised guard regions are ever read back.
- Fault injection writes through views of the owner, so it can cross the payload's logical boundary but never leaves the owning allocation; `overlap` selects whether the store ends at / starts at the boundary or straddles it, and both the guard size and the payload size are checked before the write.
- Explicit rejections for non-contiguous, nonzero-storage-offset, larger-storage-alias and non-strided templates.
- Real integration: `flashinfer.silu_and_mul(x, out=buffer.payload, enable_pdl=False)` (`flashinfer/activation.py`) — an existing `out=` injection point, no production API change; exercised at `hidden=512` (16-byte vector path) and `hidden=3420` (8-byte aligned y-half, 4-element fallback).
- Stream/graph/lifetime: reset → kernel → verify has an explicit happens-before; the graph case allocates and captures at fixed addresses, resets outside the graph, and never resets after corruption to make verification pass.

## Validation

FlashInfer base: `dc04f50c9aa3eabcdaa5feb0934edb3d85e9529a`
Hardware and software: NVIDIA L20 (SM89), driver 595.91.07, CUDA 12.8, torch 2.13.0+cu130 (CUDA 13 runtime), Python 3.10

```
CUDA_VISIBLE_DEVICES=<n> python -m pytest -q tests/utils/test_guard_zone.py tests/utils/test_guard_zone_kernel.py
# 73 passed, 2 skipped

CUDA_VISIBLE_DEVICES=<n>,<m> python -m pytest -q tests/utils/test_guard_zone.py
# 69 passed, 0 skipped   (the two per-device / cross-device attribution cases)
```

Actual execution: 73 passed / 0 failed / 2 skipped on one device; the two skipped cases pass when two devices are visible.

Case coverage: A3-01 normal payload write leaves the guards unchanged; A3-02/A3-03 prefix and suffix injection report the exact region and offset, including a boundary-straddling store; A3-04 payload correct while a guard is corrupted fails independently of `allclose`, on both the helper and the real kernel; A3-05 multi-dtype including a NaN payload; A3-06 odd shapes and alignments; A3-07 unsupported stride/alias rejection with a positive control; A3-08 half the corruption enqueued on a side stream, so verification cannot jump ahead; A3-09 graph replay keeps addresses stable and never clears corruption before verification; A3-10 per-device buffers plus same-process cross-device attribution; A3-11 allocation/release cycles without leaks or dangling payloads; A3-12 the real `silu_and_mul` output buffer.

## Results and limitations

Correctness: all 12 cases above pass on the L20.
Performance: the helper only exists in tests; `verify_guards()` synchronises the device, so it must never sit in a timing region or a graph capture.
Memory: no device-memory overhead outside the guarded allocations themselves.
Limitations: unchanged guards only prove that *those guarded bytes* were not written. Writes far from the payload, writes of the value already present, and write-then-restore are invisible; an underflow smaller than the alignment pad is not observed; this detects nothing about arbitrary cross-device corruption. It complements, and does not replace, `compute-sanitizer --mode memcheck`.

## Summary by CodeRabbit

- **Tests**
  - Added device-memory guard-zone utilities for detecting out-of-bounds writes around kernel output buffers.
  - Added validation for guard alignment, sentinel integrity, buffer lifecycle, CUDA synchronization, and unusual tensor layouts.
  - Added integration coverage confirming guarded FlashInfer activation outputs remain numerically correct while boundary overwrites are detected.

### CODE REVIEW

File: tests/test_helpers/guard_zone.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    payload_start = guard_bytes + (-(owner.data_ptr() + guard_bytes) % alignment)
+    payload_end = payload_start + payload_nbytes
+    payload = owner[payload_start:payload_end].view(dtype).view(shape)
+    assert payload.data_ptr() == owner.data_ptr() + payload_start
+    assert payload.data_ptr() % alignment == 0
+
+    buffer = GuardedBuffer(
+        owner=owner,
+        payload=payload,
+        prefix_guard=owner[:guard_bytes],
```
[the remark this answers is not in the document]
Comment (0z5a) [reply tests/test_helpers/guard_zone.py]: Fixed in 5d08ae96. The prefix guard is now the `guard_bytes` immediately before the payload (`owner[payload_start - guard_bytes : payload_start]`), so the alignment padding that used to sit unscanned between the two is gone: a store at `payload_start - 1` lands in the guard. `_scan` reports the prefix region's `owner_offset` from `payload_start - guard_bytes` as well, the module docstring shows the corrected layout, and the layout test now asserts guard-adjacency instead of `prefix_guard.data_ptr() == owner.data_ptr()`. The new A3-13 case writes exactly that byte and requires the corruption to be reported (verified on CPU and CUDA).

File: tests/test_helpers/guard_zone.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    # ``alignment`` extra bytes absorb the padding needed to align the payload
+    # at whatever address the allocator hands back; the remainder past the
+    # suffix guard is slack of the same size, left at the 0x00 fill.
+    device = torch.device(device)
+    owner = torch.zeros(
+        2 * guard_bytes + alignment + payload_nbytes, dtype=torch.uint8, device=device
+    )
+    payload_start = guard_bytes + (-(owner.data_ptr() + guard_bytes) % alignment)
+    payload_end = payload_start + payload_nbytes
+    payload = owner[payload_start:payload_end].view(dtype).view(shape)
```
[the remark this answers is not in the document]
Comment (0z5a) [reply tests/test_helpers/guard_zone.py:234]: Fixed in 5d08ae96. The payload offset step is now `lcm(alignment, dtype.itemsize)` (`tests/test_helpers/guard_zone.py:224`), so `payload_start` satisfies the requested address alignment *and* keeps the byte offset divisible by the element size -- an alignment below `itemsize` is neither rejected nor left misaligned. `owner` reserves `2 * guard_bytes + step + payload_nbytes`, which covers the larger padding. Verified with real torch over 7 alignments x 4 dtypes x 5 guard_bytes on CPU and 4 x 3 x 3 on CUDA, including the `alignment=1, guard_bytes=7, float32` case from the report (new A3-14).

**Comment by 0z5a** [comment]
The change is complete on this head and `pre-commit` is green (including the `mypy`/`ruff` hooks). The only red check left is `Report unauthorized PR status`: the GPU test lane is gated for fork PRs, so it needs a maintainer to authorize the workflow (or apply the appropriate label) before `PR Test` can run. Ready for review.

**Comment by 0z5a** [comment]
Final verification against this head, on an RTX 5090 box (the L20 is unreachable):

- `tests/utils/test_guard_zone.py` — 71 passed / 2 skipped with one device, and **73 passed / 0 skipped with two devices visible** (the two skips were exactly the per-device / cross-device attribution cases).
- `tests/utils/test_guard_zone_kernel.py`, which drives the real `silu_and_mul` — **6 passed / 0 failed** with this revision's `flashinfer/jit/activation.py` and `include/flashinfer/activation.cuh` overlaid on the pinned `flashinfer-python==0.6.18` install. That includes `hidden=3420`, the case the released vector-width logic faults on; the overlay runner (`ws/A3/run_kernel_suite_overlay.py` in the L20 package) backs up and restores both files and verifies the restore by hash.

So the three cases the change targets are pinned by a green suite on real hardware, not just by the CPU-level assertions.
